README.md
About SolrSearch 
----------

**SolrSearch** is an Omeka plugin that allows you to leverage the powerful
[Solr][solr] search engine within Omeka. Not only does Solr provide robust,
configurable, full-text indexing, it also allows you a flexible
interface to configure how your users discover the content in your
Omeka application. 

## Requirements

**SolrSearch** relies on access to a [Solr 3.5+][solr] server to
maintain the search indexes. Installation of this software is covered in
the [Solr Documentation][2]. 

### Configuration
Once Solr is up-and-running, you will need to tell Solr about the the
Omeka SolrSearch configuration. While there are many ways to define the
`solr/home` (where the index and configuration files are located),
one of the easiest ways to deal with this is by deploying Solr with a
Context which defines the path to where the `SolrSearch` plugin 
directory is located (specifically the `SolrSearch/solr-home`
directory).

The following is an example of a context file that can be deployed
easily through the [Tomcat Manager][tomcatmanager]:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Context docBase="/path/to/solr.war" debug="0" crossContext="true" >
	<Environment name="solr/home" type="java.lang.String" value="/path/to/Omeka/plugins/SolrSearch/solr-home" override="true" />
</Context>
```

It is worth noting that the `solr-home` directory can be placed
anywhere on the server that makes sense from a maitenance perspective,
which is valuable for institutions using the [multicore][multicore]
feature in Solr.

Installation and Configuration
----------
* Upload the 'SolrSearch' plugin directory to your Omeka installation's `plugin` directory. See [Installing a Plugin][plugininstall].
* Activate the plugin from the admin → Settings → Plugins page.
* Configure your connection to the Solr server. We have provided some
  typical default settings, but double check to ensure that these
settings are correct for you installation of Solr.

## Index Configuration

SolrSearch comes with some pre-selected defaults, including titles,
itemtype, tag, and references to files. These are configurable in the
Solr Index tab of Omeka's Admin interface:

TODO: add image

The index configuration is split in to the different types of
information Omeka uses, plus a few 'special' fields (image, tag,
collection, itemtype) that are outside of the normal classification.
Each element has three options:

* **Is Searchable**: Adds the text of the field to the search index.
* **Is Facet**: Adds the field as a 'facet' in the search interface
* **Is Sortable**: ??? Are we removing this?

After you have configured the fields you want indexed, and how you want
them indexed, click on the `Save Facets` button. 

**Note:** SolrSearch indexes any item marked 'public' in Omeka. 

## Hit Highlighting
SolrSearch uses 'hit highlighting' to contextualize the query result.
You can configure this in the `Hit Highlighting Options` tab.

## Reindexing


# Developer Mode
There are a number of technologies used in the development mode for this
plugin. You will need [node][node], [ruby gems][gems], as well as
several gems (installed via [bundler][bundler]), and I recommend
[rvm][rvm].

Install the solr keg from [homebrew][homebrew]. You can then start an
instance of Solr with

```bash
$ solr path/to/SolrSearch/solr-home
```

From scratch, assuming [homebrew][homebrew] is installed.

```bash
#! /bin/bash
$ brew install node solr
$ rvm gemset create solrsearch
$ solr path/to/SolrSearch/solr-core
```

## Tomcat

If you want to run this in the Tomcat servlet container, there is a
slightly different method of configuring the application.

Install start Tomcat:

```bash
$ brew install tomcat
```

We're going to need to refer to a directory inside where brew installed
tomcat, and we'll need to pass it to some of the scripts that we run
later. Let's find out what it is and save it.

```bash
$ CATALINA_HOME=$(ls -d /usr/local/Cellar/tomcat/* | head -1)/libexec
$ export CATALINA_HOME
$ echo $CATALINA_HOME
/usr/local/Cellar/tomcat/7.0.50/libexec
```

The output of the `echo` command on the last line may have a different
version number, but everything else should look the same.


Now enable the manager application. To do this, edit
`/usr/local/Cellar/tomcat/[version]/libexec/conf/tomcat-users.xml`.
The first part of the path should be the same as the output of the `echo`
command above. Inside the `<tomcat-users>` element, add something along
the following:

```xml
<role rolename="manager-gui"/>
<user username="tomcat" password="s3cret" roles="manager-gui"/>
```

You need to have [Solr](http://lucene.apache.org/solr/) downloaded somewhere on
your computer.

You need to copy some files that are shipped with Solr to get included
in your Tomcat `$PATH`. These are the files in the `examples/lib/ext/`
directory. You will have something that looks like this:


```bash
cp path/to/solr/download/examples/lib/ext/*.jar $CATALINA_HOME/lib/
```

Now start Tomcat with the catalina shell:

```bash
$ catalina start
```
This will run Tomcat as a background process on port `8080`, which you
can access at `http://localhost:8080`. 

For Tomcat, it's easiest to pass the various values that Solr needs in
an XML configuration file. Name it `tomcat-config.xml` in your project
directory, and have it contain something like this:

```xml
<Context path="/solr" docBase="/Users/[username]/Downloads/solr-4.6.0/dist/solr-4.6.0.war" debug="0" crossContext="true">
  <Environment name="solr/home" type="java.lang.String" value="/Users/[username]/projects/SolrSearch/solr-core" override="true"/>
</Context>
```

**Note:** Adjust your paths as necessary.

Now to access the manager application. Point your browser at
`http://localhost:8080/manager/html`. In the deploy section fill out the
**Context** and **XML Configuration file URL** fields with:

* **Context** /solr
* **XML Configuration file URL** path/to/tomcat-config.xml

Then hit the deploy button. If everything went well, you should see the
Solr admin panel when you point your browser at
`http://localhost:8080/solr`

## Dependencies

We use an assortment of tools in our development cycle to automate
various tasks. 

* [Rake][rake]: An Make-like program implemented in Ruby. Runs
  [Jasmine][jasmin] BDD JavaScript tests 
* [Cake][cake]: `cake` is a simplified version of Make (Rake, etc.) for
  CoffeeScript. This is used to compile JavaScript files, walk and
mangle the AST, and generate an optimized version of the JavaScript
using [uglify-js][uglify].
* [Ant][ant]: A Java tool and library used to build software. 
* [Guard][guard]: A tool that monitors file-system modification. When a
  file in this project changes, Guard automagically rebuilds compiled
sources and refreshes web pages. 
* [Pear][pear]: A framework and distribution system for reusable PHP components. We use various PHP libraries to run reports, test our code, and generate documentation. 
* [SCSS][sass]: SCSS is a syntax of SASS (syntactically awesome
  stylesheets). We use this to simplify writing stylesheets for our
plugins
* [Compass][compass]: We use this to compile our SCSS, as well as add in
  mixins for CSS3 support. 

### Guard + Compass + LiveReload + uglify + rvm
We use [RVM][rvm] to manage Ruby version, and there is an `.rvmrc` file
to automatically switch bundles and make sure all of the require Ruby
gems are install. If you do not have RVM installed, you can do all of
this manually:

```bash
cd /path/to/SolrSearch
bundle
```

This will install all of the required gems, including the guard gem to
monitor all of the files in the plugin. To start guard, simply run it
within `bundle exec`.

```bash
cd /path/to/SolrSearch
bundle exec guard
```

### Running Tests
SolrSearch tests are designed to run anywhere on the system, as long as 
an environmental variable is set that points to the target Omeka
install. This is especially useful in testing plugins against several
versions of Omeka.

In Bash, you can set this in your .bash_profile/.bash_rc file:

```bash
export OMEKA_DIR=/path/to/omeka
```

With this set, you can run the `phpunit` tests in tests:

```bash
cd /path/to/omeka/tests
phpunit
```

This will generate a test coverage report in the build sub-directory.

### Reports
This plugin uses various tools to analyze software quality from PEAR
including:

* [PHP_Depend][pdepend]: Static code analysis for PHP
* [phpDocumentor][phpdoc]: Generates documentation from PHP source code
* [PHPMessDetector][phpmd]: Code quality analysis
* [phpcpd][phpcpd]: Copy/Paste Detector (CPD) for PHP code
* [PHP_CodeSniffer][phpcs]: Tokenises PHP, JavaScript and CSS files and detects violations of a defined set of coding standards
* [PHPUnit][phpunit]: Automated testing for tests
* [PHP_CodeBrowser][phpcb]: Generates a browsable representation of PHP code where sections with violations found by quality assurance tools

### Packaging

There are two tasks for packaging the plugin. An arbitrary zip (and
tarball) can be generated with the version number and a timestamp with
the `ant zip` task.

```bash
$ ant zip
Buildfile: /path/to/SolrSearch/build.xml

zip:
      [zip] Building zip: /path/to/SolrSearch/build/dist/SolrSearch-20120618-1612.zip

BUILD SUCCESSFUL
Total time: 1 second
```

Likewise, a package suitable for uploading the the Omeka plugin
repository may be generated with the `ant package` task.

```bash
$ ant package
Buildfile: /path/to/SolrSearch/build.xml

clean:
   [delete] Deleting directory /path/to/SolrSearch/build

package:
    [mkdir] Created dir: /path/to/SolrSearch/build/dist
      [zip] Building zip: /path/to/SolrSearch/build/dist/SolrSearch-1.0.zip

BUILD SUCCESSFUL
Total time: 1 second
```

# Reporting Bugs and Contributing
If you discover a problem with SolrSearch, we would like to know about it. However, we ask that you please review these guidelines before submitting a bug report: [Bug reports][bugs].

We hope that you will consider contributing to SolrSearch. Please read this short overview for some information about to to get started: [Contributing][contributing].

## Translations

We'd welcome any help we can get translating text in the SolrSearch plugin.
We're using [Transifex][transifex] to manage translations. For more information
about using this, see [Jeremy's excellent blog post][i18nblog] about it.

[bugs]: https://github.com/scholarslab/SolrSearch/wiki/Bug-Reports
[contributing]: https://github.com/scholarslab/SolrSearch/wiki/Contributing

[phpcb]: https://github.com/Mayflower/PHP_CodeBrowser

[pdepend]: http://pdepend.org/
[phpdoc]: http://www.phpdoc.org/
[phpmd]: http://phpmd.org/
[phpcs]: http://pear.php.net/package/PHP_CodeSniffer/
[phpunit]: http://www.phpunit.de/manual/current/en/index.html
[phpcpd]: https://github.com/sebastianbergmann/phpcpd

[1]: http://scholarslab.org/ "http://scholarslab.org/"
[2]: http://lucene.apache.org/solr/#getstarted "http://lucene.apache.org/solr/#getstarted"
[3]: https://github.com/scholarslab/SolrSearch "https://github.com/scholarslab/SolrSearch"
[4]: http://github.com/scholarslab/SolrSearch/tarball/master "http://github.com/scholarslab/SolrSearch/tarball/master"
[5]: /codex/Installing_a_Plugin "Installing a Plugin"
[homebrew]: http://mxcl.github.com/homebrew/

[node]: http://nodejs.org/
[gems]: http://rubygems.org/
[bundler]: http://gembundler.com/
[rvm]: http://beginrescueend.com/
[solr]: http://lucene.apache.org/solr
[solrinstall]: http://wiki.apache.org/solr/SolrInstall
[tomcatmanager]: http://tomcat.apache.org/tomcat-6.0-doc/manager-howto.html
[multicore]: http://wiki.apache.org/solr/CoreAdmin
[rake]: http://rubygems.org/gems/rake
[cake]: http://coffeescript.org/documentation/docs/cake.html
[jasmin]: http://pivotal.github.com/jasmine/
[uglify]: https://github.com/mishoo/UglifyJS
[ant]: http://ant.apache.org/
[guard]: https://github.com/guard/guard
[pear]: http://pear.php.net/
[sass]: http://sass-lang.com/
[compass]: http://compass-style.org/
[rvm]: https://rvm.io/

[transifex]: http://www.transifex.com/
[i18nblog]: http://www.scholarslab.org/slab-code/translating-neatline/


CHANGELOG
[Thu, 7 Jun 2012 10:00:41 -0400] Display the result type in the
search results. Fixes #60.

[Wed, 6 Jun 2012 16:20:46 -0400] Check that db tables exist before
trying to index their data. Fixes #63.

[Wed, 6 Jun 2012 15:10:07 -0400] Added a select-all checkbox at the
head of each group/column. Fixes #57.

[Wed, 6 Jun 2012 10:41:59 -0400] Merge branch 'feature/rename-facet'
into develop. Fixes #55.

[Tue, 5 Jun 2012 15:48:33 -0400] Merge branch 'feature/highlighting'
into develop

[Mon, 4 Jun 2012 15:50:32 -0400] Merge branch 'feature/reindex'
into develop

[Wed, 30 May 2012 09:48:54 -0400] minor formatting change

[Wed, 30 May 2012 09:47:17 -0400] Hides the configuration option
to set the number of search results in the Solr response object.
This will be inferred from the global Omeka results per page option
in the admin settings, and asynchronously send to the client with
infiniate scroll. Closes #54

[Wed, 30 May 2012 09:37:38 -0400] Adds a check if the admin theme
is in place, and doesn't override the simple-search uri. Since
SolrSearch only indexes items marked 'public,' this effectively
broke the searching mechanism in the admin panel for items. Closes
#58

[Wed, 30 May 2012 09:16:03 -0400] Updates the labels to be more
descriptive of the elements in defining hit highlighting, including
field descriptions in plain terms; closes 56

[Thu, 17 May 2012 13:56:46 -0400] Updates the configuration error
to inform user that the plugin cannot ping the Solr server.

[Thu, 17 May 2012 13:09:03 -0400] Added a facet label field to the
facet table and include spaces in the standard facet labels. Fixes
#51.

[Thu, 17 May 2012 10:51:24 -0400] Create an absolute URL from what's
stored in Solr. Fixes #50.

[Thu, 17 May 2012 09:58:40 -0400] Make sure to retrieve all hidden
fields and have something for untitled docs. Fixes #49.

[Wed, 16 May 2012 16:56:44 -0400] minifies changes to the scroll
js

[Wed, 16 May 2012 16:54:01 -0400] remove link to previous page from
pagination; wraps msgText and finishedMsg for infinite scroll;
closes #47

[Wed, 16 May 2012 16:17:43 -0400] Improved tests for null queries.

[Tue, 15 May 2012 15:32:37 -0400] Merge branch 'feature/search-addons'
into develop. Fixes #8.

[Mon, 14 May 2012 13:54:16 -0300] Remove errant A in file.

[Mon, 14 May 2012 12:10:05 -0400] ignore the build directory

[Mon, 14 May 2012 12:08:19 -0400] Update the test runners to use a
bootstrap

[Sun, 13 May 2012 08:07:22 -0400] Adds a field lookup to
SolrSearch_Helpers_View::lookupElement until this is stored in the
index

[Thu, 10 May 2012 11:41:03 -0400] Updates markup and styles for
results page.

[Thu, 10 May 2012 11:41:03 -0400] Hides #solr-nav div.

[Thu, 10 May 2012 11:41:03 -0400] Moves all markup surrounding
pagination to the partial.

[Thu, 10 May 2012 10:50:18 -0400] coffee-script provides cake

[Thu, 10 May 2012 10:48:33 -0400] Merge branch 'develop' of
github.com:scholarslab/SolrSearch into develop

[Thu, 10 May 2012 10:48:24 -0400] Updates node dependencies for
uglify and cake

[Thu, 10 May 2012 08:21:08 -0400] Renames Solr Index admin tab to
Solr Search; cloases #41

[Thu, 10 May 2012 07:33:44 -0400] Merge branch 'feature/infinitescroll'
into develop

[Sun, 6 May 2012 09:16:01 -0300] Fixes typo in compass config file
to the images directory

[Thu, 3 May 2012 10:11:35 -0400] Pulls guards out of the frontend
group

[Thu, 3 May 2012 09:27:39 -0400] Merge branch 'develop' of
github.com:scholarslab/SolrSearch into develop

[Thu, 3 May 2012 09:27:21 -0400] Updates the labels in the admin
section; closes #43

[Wed, 2 May 2012 15:01:07 -0400] Centers headings and checkboxes
for index config form.

[Wed, 2 May 2012 14:44:14 -0400] Merge branch 'feature/sort-removal'
into develop. Fixes #25.

[Wed, 2 May 2012 14:26:18 -0400] Bedazzles indexing configuration
tables in admin.

[Wed, 2 May 2012 11:05:17 -0400] Adds zebrastripe to make checkbox
a bit easier to find; #42

[Wed, 2 May 2012 10:53:07 -0400] Updates the Readme for the new
version of the plugin

[Tue, 1 May 2012 20:14:29 -0400] Clean up deprecated elements from
the Solr schema

[Tue, 1 May 2012 20:12:03 -0400] Updates the solrconfig to work in
solr 3.3+ for the admin ping (referenced #36); adds the default
field in to the search and select request handlers (closes #40);
adds the default operator of AND for the query op (closes #39);
fixes ping and browse excpetions (closes #26)

[Tue, 1 May 2012 17:00:00 -0400] Merge branch 'develop' of
github.com:scholarslab/SolrSearch into develop

[Tue, 1 May 2012 16:58:40 -0400] Always include the image in the
list of fields requested of Solr. Fixes #28.

[Tue, 1 May 2012 16:57:23 -0400] Always index the image IDs so they
can be included in the output.

[Tue, 1 May 2012 16:57:01 -0400] Move the image above the tags in
the output.

[Tue, 1 May 2012 16:24:44 -0400] By default, also index dc:title
and dc:description. Fixes #37.

[Tue, 1 May 2012 15:49:04 -0400] Merge branch 'develop' of
github.com:scholarslab/SolrSearch into develop

[Tue, 1 May 2012 15:48:56 -0400] Properly look up field names from
the database in the public view; temporarily fixes #35, but this
should really be taken care of when the element is indexed, not
re-rended to the view

[Thu, 26 Apr 2012 21:28:27 -0400] Don't display facet categories
that don't have any hits displayed.

[Thu, 26 Apr 2012 15:08:52 -0400] Actually index tags.

[Thu, 26 Apr 2012 14:57:25 -0400] Sort the facet listing.

[Thu, 26 Apr 2012 14:53:39 -0400] Use a hand pointer cursor over
the facet category labels.

[Thu, 26 Apr 2012 14:50:56 -0400] Use lower-case facet value labels
so they get connected with what Solr uses.

[Thu, 26 Apr 2012 14:32:49 -0400] Make sure to save all results.
Fixes #33.

[Thu, 26 Apr 2012 14:06:46 -0400] Removed FedoraConnector special
handling. Fixes #34.

[Thu, 26 Apr 2012 14:01:16 -0400] Only index fields that need to
be indexed.

[Thu, 26 Apr 2012 12:35:10 -0400] Clarified facet configuration
form labels. Fixes #30.


SolrPlugin.php
<?php
/* vim: set expandtab tabstop=4 shiftwidth=4 softtabstop=4: */

/**
 * SolrSearch Omeka plugin
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at http://www.apache.org/licenses/LICENSE-2.0 Unless required by
 * applicable law or agreed to in writing, software distributed under the
 * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS
 * OF ANY KIND, either express or implied. See the License for the specific
 * language governing permissions and limitations under the License.
 *
 * @package omeka
 * @subpackage SolrSearch
 * @author Scholars' Lab
 * @license http://www.apache.org/licenses/LICENSE-2.0 Apache 2.0
 * @copyright 2010 The Board and Visitors of the University of Virginia
 * @link https://github.com/scholarslab/SolrSearch/
 *
 * PHP version 5
 */

class SolrPlugin
{
    // {{{ hooks
    private static $_hooks = array(
        'install',
        'uninstall',
        'before_delete_item',
        'after_save_item',
        'define_routes',
        'define_acl',
        'admin_theme_header',
        'public_theme_header',
        'config_form',
        'config'
    );
    //}}}

    //{{{ filters
    private static $_filters = array(
        'admin_navigation_main',
        'simple_search_default_uri'
    );
    //}}}

    public function __construct()
    {
        $this->_db = get_db();
        self::addHooksAndFilters();
    }

    public function addHooksAndFilters()
    {
        foreach(self::$_hooks as $hookName) {
            $functionName = Inflector::variablize($hookName);
            add_plugin_hook($hookName, array($this, $functionName));
        }

        foreach(self::$_filters as $filterName) {
            $functionName = Inflector::variablize($filterName);
            add_filter($filterName, array($this, $functionName));
        }
    }

    public function install()
    {
        self::_createSolrTable();
        self::_addFacetMappings();
        self::_setOptions();
    }


    public function uninstall()
    {
        $sql = "DROP TABLE IF EXISTS `{$this->_db->prefix}solr_search_facets`";
        $this->_db->query($sql);

        $solr = new Apache_Solr_Service(SOLR_SERVER, SOLR_PORT, SOLR_CORE);
        $solr->deleteByQuery('*:*');
        $solr->commit();
        $solr->optimize();

        self::_deleteOptions();
    }

    public function beforeDeleteItem($item)
    {
        $solr = new Apache_Solr_Service(SOLR_SERVER, SOLR_PORT, SOLR_CORE);
        $solr->deleteByQuery('id:' . $item['id']);
        $solr->commit();
        $solr->optimize();
    }

    public function afterSaveItem($item)
    {
        $solr = new Apache_Solr_Service(SOLR_SERVER, SOLR_PORT, SOLR_CORE);

        if ($item['public'] == true) {
            $docs = array();
            $doc = SolrSearch_IndexHelpers::itemToDocument($this->_db, $item);
            $docs[] = $doc;

            $solr->addDocuments($docs);
            $solr->commit();
            $solr->optimize();
        } else {
            // If the item's no longer public, remove it from the index.
            $solr->deleteByQuery('id:' . $item['id']);
            $solr->commit();
            $solr->optimize();
        }
    }

    public function defineRoutes($router)
    {
        $searchResultsRoute = new Zend_Controller_Router_Route(
            'results', array(
                'controller' => 'search',
                'action'     => 'results',
                'module'     => 'solr-search'
            ));
        $router->addRoute('solr_search_results_route', $searchResultsRoute);
    }

    public function defineAcl($acl)
    {
        $acl->loadResourceList(array(
            'SolrSearch_Config' => array('index', 'status')
        ));
    }

    public function adminThemeHeader($request)
    {
        $module = $request->getModuleName();
        if ($module == 'solr-search' || $module == 'default') {
            queue_css('solr_search');
        }
    }

    public function publicThemeHeader()
    {
        queue_css('solr_search');
        $js = 'solrsearch-' . SOLR_SEARCH_PLUGIN_VERSION . '-min';
        queue_js($js);
    }

    public function configForm()
    {
        $fields = $this->_makeConfigFields();
        $buffer = array();

        $buffer[] = '<div class="field solrsearch config"><h3>Solr Options</h3>';
        foreach ($fields as $field) {
            $buffer[] = $field->render();
        }
        $buffer[] = '</div>';

        echo join($buffer);
    }

    public function config()
    {
        $form = $this->_makeConfigForm();

        if ($form->isValid($_POST)) {
            $options = $form->getValues();

            if (!SolrSearch_IndexHelpers::pingSolrServer($options)) {
                throw new Omeka_Validator_Exception(
                    "Invalid Solr server host, port, or core."
                );
            }

            foreach ($options as $option => $value) {
                set_option($option, $value);
            }

            $this->_flashSuccess("Config updated.");

            ProcessDispatcher::startProcess('SolrSearch_IndexAll', null, $args);
        } else {
            $output = '';
            foreach ($form->getMessages() as $code => $msgs) {
                foreach ($msgs as $msg) {
                    $output .= "$msg ($code)\n";
                }
            }
            throw new Omeka_Validator_Exception($output);
        }
    }

    public function adminNavigationMain($tabs)
    {
        if (get_acl()->checkUserPermission('SolrSearch_Config', 'index')) {
            $tabs['Solr Index'] = uri('solr-search/config/');
        }
        return $tabs;
    }

    public function simpleSearchDefaultUri($uri)
    {
        $uri = uri('solr-search/results/interceptor');
        return $uri;
    }

    // {{{protected

    /**
     * Populates the facet table with human readable mappings of Omeka Element ids
     *
     * There are special cases for sorting <tt>tags</tt>,
     * <tt>collection</tt>, and <tt>itemType</tt>
     *
     */
    protected function _addFacetMappings()
    {
        $elements = $this->_db->getTable('Element')->findAll();
        $sql = <<<SQL
            INSERT INTO `{$this->_db->prefix}solr_search_facets`
                (element_id, name, element_set_id, is_facet, is_displayed, is_sortable)
                VALUES (?, ?, ?, ?, ?, ?);
SQL;
        $stmt = $this->_db->prepare($sql);

        $stmt->execute(array(null, 'image',      null, 1, 1, 1));
        $stmt->execute(array(null, 'tag',        null, 1, 1, 1));
        $stmt->execute(array(null, 'collection', null, 1, 1, 1));
        $stmt->execute(array(null, 'itemtype',   null, 1, 1, 1));

        foreach ($elements as $element) {
            $stmt->execute(array(
                $element['id'], $element['name'], $element['element_set_id'], 0, 0, 0
            ));
        }
    }

    protected function _setOptions()
    {
        set_option('solr_search_server', 'localhost');
        set_option('solr_search_port', '8080');
        set_option('solr_search_core', '/solr/');
        set_option('solr_search_rows', '');
        set_option('solr_search_facet_limit', '25');
        set_option('solr_search_hl', 'true');
        set_option('solr_search_snippets', '1');
        set_option('solr_search_fragsize', '250');
        set_option('solr_search_facet_sort', 'count');
    }

    protected function _deleteOptions()
    {
        delete_option('solr_search_server');
        delete_option('solr_search_port');
        delete_option('solr_search_core');
        delete_option('solr_search_rows');
        delete_option('solr_search_facet_limit');
        delete_option('solr_search_hl');
        delete_option('solr_search_snippets');
        delete_option('solr_search_fragsize');
        delete_option('solr_search_facet_sort');
    }


    protected function _createSolrTable()
    {
        $sql = <<<SQL
      CREATE TABLE IF NOT EXISTS `{$this->_db->prefix}solr_search_facets` (
        `id` int(10) unsigned NOT NULL auto_increment,
            `element_id` int(10) unsigned,
            `name` tinytext collate utf8_unicode_ci NOT NULL,
            `element_set_id` int(10) unsigned,
            `is_facet` tinyint unsigned DEFAULT 0,
            `is_displayed` tinyint unsigned DEFAULT 0,
            `is_sortable` tinyint unsigned DEFAULT 0,
        PRIMARY KEY  (`id`)
       ) ENGINE=MyISAM DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;
SQL;

        $this->_db->exec($sql);

    }

    protected function _makeConfigForm() {
        $form = new Zend_Form();
        $this->_makeConfigFields($form);
        return $form;
    }

    protected function _makeConfigFields($form=null) {
        $fields = array();

        $fields[] = $this->_makeOptionField(
            $form, 'solr_search_server', 'Server Host:', true
        );
        $fields[] = $this->_makeOptionField(
            $form, 'solr_search_port', 'Server Port:', true
        )
            ->addValidator(new Zend_Validate_Digits());
        $fields[] = $this->_makeOptionField(
            $form, 'solr_search_core', 'Solr Core Name:', true
        )
            ->addValidator('regex', true, array('/\/.*\//i'));

        $fields[] = $this->_makeOptionField(
            $form, 'solr_search_rows', "Results Per Page\n(empty uses Omeka's paging settings):", false
        )
            ->addValidator(new Zend_Validate_Digits())
            ->addErrorMessage('Results count must be numeric');

        $fields[] = $this->_makeOptionField(
            $form, 'solr_search_facet_sort', 'Default Sort Order:', false,
            'Zend_Form_Element_Select'
        )
            ->addMultiOption('index', 'Alphabetical')
            ->addMultiOption('count', 'Occurrences');

        $fields[] = $this->_makeOptionField(
            $form, 'solr_search_facet_limit', 'Maximum Facet Count:', true
        )
            ->addValidator(new Zend_Validate_Digits());

        return $fields;
    }

    protected function _makeOptionField(
        $form, $name, $label, $required, $cls='Zend_Form_Element_Text'
    ) {
        $field = new $cls($name, array(
            'label'    => $label,
            'value'    => get_option($name),
            'required' => $required
        ));

        if ($form != null) {
            $form->addElement($field);
        }

        return $field;
    }

    /**
     * This sets a flash error message.
     *
     * @param string $msg The message to flash.
     *
     * @return void
     * @author Eric Rochester <erochest@virginia.edu>
     **/
    protected function _flashError($msg)
    {
        $flash = new Omeka_Controller_Flash;
        $flash->setFlash(Omeka_Controller_Flash::GENERAL_ERROR, 
                         $msg, 
                         Omeka_Controller_Flash::DISPLAY_NEXT);
    }

    protected function _flashSuccess($msg)
    {
        $flash = new Omeka_Controller_Flash;
        $flash->setFlash(Omeka_Controller_Flash::SUCCESS, 
                         $msg, 
                         Omeka_Controller_Flash::DISPLAY_NEXT);
    }
    //}}}

}

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * c-hanging-comment-ender-p: nil
 * End:
 */
